src/fondant/pipeline/lightweight_component.py

```diff
@@ -264,6 +263,39 @@ def consumes(cls) -> t.Optional[t.Dict[str, t.Dict[t.Any, t.Any]]]:
     return wrapper
 
 
+def new_getfile(_object, _old_getfile=inspect.getfile):
+    if not inspect.isclass(_object):
+        return _old_getfile(_object)
+
+    # Lookup by parent module (as in current inspect)
+    if hasattr(_object, "__module__"):
+        object_ = sys.modules.get(_object.__module__)
+        if hasattr(object_, "__file__"):
+            return object_.__file__
+
+    # If parent module is __main__, lookup by methods (NEW)
+    for name, member in inspect.getmembers(_object):
+        if (
+            inspect.isfunction(member)
+            and _object.__qualname__ + "." + member.__name__ == member.__qualname__
+        ):
+            return inspect.getfile(member)
+
+    msg = f"Source for {_object!r} not found"
+    raise TypeError(msg)
+
+
+def is_running_in_notebook():
+    """Check if the code is running in a Jupyter notebook."""
+    try:
+        from IPython import get_ipython
+
+        shell = get_ipython().__class__.__name__
+        return shell == "ZMQInteractiveShell"
```

Comment:
What about running in an IPython terminal instead of a notebook? (if `shell == "TerminalInteractiveShell"`)

Reply:
Does this fix work for the other shells as well?

Reply:
I tested on my IDE python consol shell (`PyDevTerminalInteractiveShell`), jupter notebook (`ZMQInteractiveShell`) and colab (`Shell`). I presume those are the main 3 options that will be used for compilation (even IDE console seems a bit unlikely)